File: focusling/modules/focusling-protection/ios/DetectionCore/Sources/DetectionCore/DetectionPolicy.swift
import Foundation

/// Surfaces the detector can report. Raw values match the TypeScript `ContentSurface`.
public enum ContentSurface: String, Codable, Sendable {
  case instagramReels, instagramStories, instagramExplore, youtubeShorts, other
}

public struct SurfaceClassification: Equatable, Sendable {
  public enum App: String, Sendable { case instagram, unknown }
  public let app: App
  public let surface: ContentSurface
  public let confidence: Double
  public init(app: App, surface: ContentSurface, confidence: Double) {
    self.app = app
    self.surface = surface
    self.confidence = confidence
  }
  public static let nothing = SurfaceClassification(app: .unknown, surface: .other, confidence: 0)
}

/// Configurable temporal confirmation. Values arrive from JS (`config/protection.ts`).
public struct DetectionPolicyConfig: Codable, Equatable, Sendable {
  public var confidenceThreshold: Double
  public var requiredVotes: Int
  public var windowSize: Int
  public var minimumDetectionMs: Double
  public var cooldownMs: Double

  public init(confidenceThreshold: Double = 0.75, requiredVotes: Int = 3, windowSize: Int = 5, minimumDetectionMs: Double = 1500, cooldownMs: Double = 20_000) {
    self.confidenceThreshold = confidenceThreshold
    self.requiredVotes = requiredVotes
    self.windowSize = windowSize
    self.minimumDetectionMs = minimumDetectionMs
    self.cooldownMs = cooldownMs
  }

  public init(dictionary: [String: Double]) {
    let d = DetectionPolicyConfig()
    self.init(
      confidenceThreshold: dictionary["confidenceThreshold"] ?? d.confidenceThreshold,
      requiredVotes: Int(dictionary["requiredVotes"] ?? Double(d.requiredVotes)),
      windowSize: Int(dictionary["windowSize"] ?? Double(d.windowSize)),
      minimumDetectionMs: dictionary["minimumDetectionMs"] ?? d.minimumDetectionMs,
      cooldownMs: dictionary["cooldownMs"] ?? d.cooldownMs
    )
  }
}

/// Rolling-window vote. Mirrors `src/core/protection/detectionPolicy.ts` exactly;
/// both are checked against `protectionVectors.json`.
public struct DetectionPolicy: Sendable {
  public private(set) var config: DetectionPolicyConfig
  public let target: ContentSurface
  /// Timestamp (ms) of each positive sample, or nil for a negative; oldest first.
  private var samples: [Double?] = []
  private var cooldownUntil: Double = 0

  public init(config: DetectionPolicyConfig, target: ContentSurface = .instagramReels) {
    self.config = config
    self.target = target
  }

  public var votes: Int { samples.compactMap { $0 }.count }

  public mutating func reset() {
    samples = []
    cooldownUntil = 0
  }

  /// Returns true when an intervention should happen now.
  public mutating func record(_ classification: SurfaceClassification, atMs at: Double) -> Bool {
    if at < cooldownUntil {
      samples = []
      return false
    }
    let positive = classification.surface == target && classification.confidence >= config.confidenceThreshold
    samples.append(positive ? at : nil)
    if samples.count > config.windowSize { samples.removeFirst(samples.count - config.windowSize) }

    let positives = samples.compactMap { $0 }
    let span = (positives.last ?? 0) - (positives.first ?? 0)
    if positives.count >= config.requiredVotes && span >= config.minimumDetectionMs {
      samples = []
      cooldownUntil = at + config.cooldownMs
      return true
    }
    return false
  }
}

File: focusling/src/core/protection/detectionPolicy.ts
import type { DetectionPolicyConfig } from '@/config/protection';
import type { ContentSurface, SurfaceClassification, Timestamp } from '../models';

/**
 * Rolling-window vote over detector samples. Pure and deterministic: the native
 * Swift implementation follows exactly the same rules (see protectionVectors.json).
 */
export interface DetectionState {
  /** Most recent samples, oldest first: timestamp of each positive, or null for a negative. */
  samples: (Timestamp | null)[];
  cooldownUntil: Timestamp;
}

export type DetectionDecision = { intervene: false; votes: number } | { intervene: true; votes: number; surface: ContentSurface };

export function createDetectionState(): DetectionState {
  return { samples: [], cooldownUntil: 0 };
}

export function isPositive(c: SurfaceClassification, target: ContentSurface, policy: DetectionPolicyConfig): boolean {
  return c.surface === target && c.confidence >= policy.confidenceThreshold;
}

/**
 * Feed one sample. Returns the new state and whether to intervene now.
 * Intervention requires `requiredVotes` positives within the last `windowSize`
 * samples, spanning at least `minimumDetectionMs`, and not during cooldown.
 * Intervening starts the cooldown and clears the window.
 */
export function recordSample(
  state: DetectionState,
  classification: SurfaceClassification,
  at: Timestamp,
  target: ContentSurface,
  policy: DetectionPolicyConfig,
): { state: DetectionState; decision: DetectionDecision } {
  if (at < state.cooldownUntil) {
    return { state: { ...state, samples: [] }, decision: { intervene: false, votes: 0 } };
  }
  const samples = [...state.samples, isPositive(classification, target, policy) ? at : null].slice(-policy.windowSize);
  const positives = samples.filter((s): s is number => s !== null);
  const votes = positives.length;
  const span = votes > 0 ? positives[positives.length - 1]! - positives[0]! : 0;

  if (votes >= policy.requiredVotes && span >= policy.minimumDetectionMs) {
    return {
      state: { samples: [], cooldownUntil: at + policy.cooldownMs },
      decision: { intervene: true, votes, surface: target },
    };
  }
  return { state: { ...state, samples }, decision: { intervene: false, votes } };
}
